File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/core/di/module/application/NetModule.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan.core.di.module.application;

import android.content.Context;
import android.net.ConnectivityManager;

import com.github.k1rakishou.ChanSettings;
import com.github.k1rakishou.chan.Chan;
import com.github.k1rakishou.chan.core.base.okhttp.CoilOkHttpClient;
import com.github.k1rakishou.chan.core.base.okhttp.HttpLoggingInterceptorLazy;
import com.github.k1rakishou.chan.core.base.okhttp.ProxiedOkHttpClient;
import com.github.k1rakishou.chan.core.base.okhttp.RealDownloaderOkHttpClient;
import com.github.k1rakishou.chan.core.base.okhttp.RealProxiedOkHttpClient;
import com.github.k1rakishou.chan.core.cache.CacheHandler;
import com.github.k1rakishou.chan.core.cache.FileCacheV2;
import com.github.k1rakishou.chan.core.helper.ProxyStorage;
import com.github.k1rakishou.chan.core.manager.FirewallBypassManager;
import com.github.k1rakishou.chan.core.site.SiteResolver;
import com.github.k1rakishou.chan.core.site.http.HttpCallManager;
import com.github.k1rakishou.common.AppConstants;
import com.github.k1rakishou.common.dns.DnsOverHttpsSelectorFactory;
import com.github.k1rakishou.common.dns.NormalDnsSelectorFactory;
import com.github.k1rakishou.core_logger.Logger;
import com.github.k1rakishou.fsaf.FileManager;
import com.google.gson.Gson;

import javax.inject.Singleton;

import dagger.Lazy;
import dagger.Module;
import dagger.Provides;
import kotlinx.coroutines.CoroutineScope;

@Module
public class NetModule {

    @Provides
    @Singleton
    public HttpLoggingInterceptorLazy provideHttpLoggingInterceptorLazy() {
        return new HttpLoggingInterceptorLazy();
    }

    @Provides
    @Singleton
    public ProxyStorage provideProxyStorage(
            CoroutineScope appScope,
            Context appContext,
            AppConstants appConstants,
            SiteResolver siteResolver,
            Gson gson
    ) {
        Logger.deps("ProxyStorage");
        return new ProxyStorage(
                appScope,
                appContext,
                appConstants,
                ChanSettings.verboseLogs.get(),
                siteResolver,
                gson
        );
    }

    @Provides
    @Singleton
    public CacheHandler provideCacheHandler(AppConstants appConstants) {
        Logger.deps("CacheHandler");

        return new CacheHandler(
                ChanSettings.prefetchMedia.get(),
                appConstants
        );
    }

    @Provides
    @Singleton
    public FileCacheV2 provideFileCacheV2(
            ConnectivityManager connectivityManager,
            FileManager fileManager,
            Lazy<CacheHandler> cacheHandler,
            SiteResolver siteResolver,
            Lazy<RealDownloaderOkHttpClient> realDownloaderOkHttpClient,
            AppConstants appConstants
    ) {
        Logger.deps("FileCacheV2");

        return new FileCacheV2(
                fileManager,
                cacheHandler,
                siteResolver,
                realDownloaderOkHttpClient,
                connectivityManager,
                appConstants
        );
    }

    @Provides
    @Singleton
    public HttpCallManager provideHttpCallManager(
            Lazy<ProxiedOkHttpClient> okHttpClient,
            AppConstants appConstants
    ) {
        Logger.deps("HttpCallManager");

        return new HttpCallManager(okHttpClient, appConstants);
    }

    /**
     * This okHttpClient is for posting.
     */
    @Provides
    @Singleton
    public ProxiedOkHttpClient provideProxiedOkHttpClient(
            NormalDnsSelectorFactory normalDnsSelectorFactory,
            DnsOverHttpsSelectorFactory dnsOverHttpsSelectorFactory,
            Chan.OkHttpProtocols okHttpProtocols,
            ProxyStorage proxyStorage,
            HttpLoggingInterceptorLazy httpLoggingInterceptorLazy,
            SiteResolver siteResolver,
            FirewallBypassManager firewallBypassManager
    ) {
        Logger.deps("RealProxiedOkHttpClient");

        return new RealProxiedOkHttpClient(
                normalDnsSelectorFactory,
                dnsOverHttpsSelectorFactory,
                okHttpProtocols,
                proxyStorage,
                httpLoggingInterceptorLazy,
                siteResolver,
                firewallBypassManager
        );
    }

    /**
     * This okHttpClient is for Coil image loading library
     */
    @Provides
    @Singleton
    public CoilOkHttpClient provideCoilOkHttpClient(
            Context applicationContext,
            NormalDnsSelectorFactory normalDnsSelectorFactory,
            DnsOverHttpsSelectorFactory dnsOverHttpsSelectorFactory,
            Chan.OkHttpProtocols okHttpProtocols,
            ProxyStorage proxyStorage,
            HttpLoggingInterceptorLazy httpLoggingInterceptorLazy,
            SiteResolver siteResolver,
            FirewallBypassManager firewallBypassManager
    ) {
        Logger.deps("CoilOkHttpClient");

        return new CoilOkHttpClient(
                applicationContext,
                normalDnsSelectorFactory,
                dnsOverHttpsSelectorFactory,
                okHttpProtocols,
                proxyStorage,
                httpLoggingInterceptorLazy,
                siteResolver,
                firewallBypassManager
        );
    }

    /**
     * This okHttpClient is for images/file/apk updates/ downloading, prefetching, etc.
     */
    @Provides
    @Singleton
    public RealDownloaderOkHttpClient provideDownloaderOkHttpClient(
            NormalDnsSelectorFactory normalDnsSelectorFactory,
            DnsOverHttpsSelectorFactory dnsOverHttpsSelectorFactory,
            Chan.OkHttpProtocols okHttpProtocols,
            ProxyStorage proxyStorage,
            HttpLoggingInterceptorLazy httpLoggingInterceptorLazy,
            SiteResolver siteResolver,
            FirewallBypassManager firewallBypassManager
    ) {
        Logger.deps("RealDownloaderOkHttpClient");

        return new RealDownloaderOkHttpClient(
                normalDnsSelectorFactory,
                dnsOverHttpsSelectorFactory,
                okHttpProtocols,
                proxyStorage,
                httpLoggingInterceptorLazy,
                siteResolver,
                firewallBypassManager
        );
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/core/di/module/application/NetModule.kt
package com.github.k1rakishou.chan.core.di.module.application

import android.content.Context
import android.net.ConnectivityManager
import com.github.k1rakishou.chan.core.base.okhttp.CoilOkHttpClient
import com.github.k1rakishou.chan.core.base.okhttp.DownloaderOkHttpClient
import com.github.k1rakishou.chan.core.base.okhttp.DownloaderOkHttpClientImpl
import com.github.k1rakishou.chan.core.base.okhttp.OkHttpClientForInterceptors
import com.github.k1rakishou.chan.core.base.okhttp.OkHttpClientForInterceptorsImpl
import com.github.k1rakishou.chan.core.base.okhttp.ProxiedOkHttpClient
import com.github.k1rakishou.chan.core.base.okhttp.ProxiedOkHttpClientImpl
import com.github.k1rakishou.chan.core.base.okhttp.interceptor.Chan8MoeInterceptor
import com.github.k1rakishou.chan.core.base.okhttp.interceptor.CloudFlareInterceptor
import com.github.k1rakishou.chan.core.base.okhttp.interceptor.HttpLoggingInterceptorLazy
import com.github.k1rakishou.chan.core.base.okhttp.interceptor.KurobaOkHttpInterceptor
import com.github.k1rakishou.chan.core.cache.CacheHandler
import com.github.k1rakishou.chan.core.cache.downloader.ChunkedMediaDownloader
import com.github.k1rakishou.chan.core.cache.downloader.ChunkedMediaDownloaderImpl
import com.github.k1rakishou.chan.core.helper.ProxyStorage
import com.github.k1rakishou.chan.core.manager.FirewallBypassManager
import com.github.k1rakishou.chan.core.site.SiteResolver
import com.github.k1rakishou.chan.core.site.http.HttpCallManager
import com.github.k1rakishou.common.AppConstants
import com.github.k1rakishou.common.dns.DnsOverHttpsSelectorFactory
import com.github.k1rakishou.common.dns.NormalDnsSelectorFactory
import com.github.k1rakishou.core_logger.Logger.deps
import com.github.k1rakishou.fsaf.FileManager
import com.github.k1rakishou.v2.KurobaSettings
import com.google.gson.Gson
import dagger.Lazy
import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoSet
import kotlinx.coroutines.CoroutineScope
import javax.inject.Singleton

@Module
class NetModule {
  @Provides
  @Singleton
  fun provideHttpLoggingInterceptorLazy(): HttpLoggingInterceptorLazy {
    return HttpLoggingInterceptorLazy()
  }

  @Provides
  @Singleton
  fun provideProxyStorage(
    appScope: CoroutineScope,
    appContext: Context,
    kurobaSettings: KurobaSettings,
    appConstants: AppConstants,
    siteResolver: SiteResolver,
    gson: Gson
  ): ProxyStorage {
    deps("ProxyStorage")
    return ProxyStorage(
      appScope,
      appContext,
      appConstants,
      kurobaSettings.application.verboseLogs.readBlocking(),
      siteResolver,
      gson
    )
  }

  @Provides
  @Singleton
  fun provideCacheHandler(
    appConstants: AppConstants,
    kurobaSettings: KurobaSettings,
  ): CacheHandler {
    deps("CacheHandler")

    return CacheHandler(
      appConstants,
      kurobaSettings
    )
  }

  @Provides
  @Singleton
  fun provideChunkedMediaDownloaderImpl(
    appConstants: AppConstants,
    fileManager: FileManager,
    siteResolver: SiteResolver,
    cacheHandler: Lazy<CacheHandler>,
    downloaderOkHttpClient: Lazy<DownloaderOkHttpClient>,
    connectivityManager: ConnectivityManager
  ): ChunkedMediaDownloader {
    deps("ChunkedMediaDownloader")

    return ChunkedMediaDownloaderImpl(
      appConstants,
      fileManager,
      siteResolver,
      cacheHandler,
      downloaderOkHttpClient,
      connectivityManager
    )
  }

  @Provides
  @Singleton
  fun provideHttpCallManager(
    okHttpClient: Lazy<ProxiedOkHttpClient>,
    appConstants: AppConstants
  ): HttpCallManager {
    deps("HttpCallManager")
    return HttpCallManager(okHttpClient, appConstants)
  }

  @Provides
  @IntoSet
  fun provideCloudflareInterceptor(
    siteResolver: SiteResolver,
    firewallBypassManager: FirewallBypassManager
  ): KurobaOkHttpInterceptor {
    deps("CloudFlareInterceptor")
    return CloudFlareInterceptor(
      siteResolver = siteResolver,
      firewallBypassManager = firewallBypassManager,
    )
  }

  @Provides
  @IntoSet
  fun provideChan8MoeInterceptor(
    okHttpClient: OkHttpClientForInterceptors,
    siteResolver: SiteResolver
  ): KurobaOkHttpInterceptor {
    deps("Chan8MoeInterceptor")
    return Chan8MoeInterceptor(
      okHttpClient = okHttpClient,
      siteResolver = siteResolver
    )
  }

  /**
   * This okHttpClient is for posting.
   */
  @Provides
  @Singleton
  fun provideProxiedOkHttpClient(
    kurobaSettings: KurobaSettings,
    normalDnsSelectorFactory: NormalDnsSelectorFactory,
    dnsOverHttpsSelectorFactory: DnsOverHttpsSelectorFactory,
    proxyStorage: ProxyStorage,
    httpLoggingInterceptorLazy: HttpLoggingInterceptorLazy,
    interceptors: Set<@JvmSuppressWildcards KurobaOkHttpInterceptor>
  ): ProxiedOkHttpClient {
    deps("ProxiedOkHttpClient")

    return ProxiedOkHttpClientImpl(
      kurobaSettings = kurobaSettings,
      normalDnsSelectorFactory = normalDnsSelectorFactory,
      dnsOverHttpsSelectorFactory = dnsOverHttpsSelectorFactory,
      proxyStorage = proxyStorage,
      httpLoggingInterceptorLazy = httpLoggingInterceptorLazy,
      interceptors = interceptors
    )
  }

  /**
   * This okHttpClient is for Coil image loading library
   */
  @Provides
  @Singleton
  fun provideCoilOkHttpClient(
    kurobaSettings: KurobaSettings,
    normalDnsSelectorFactory: NormalDnsSelectorFactory,
    dnsOverHttpsSelectorFactory: DnsOverHttpsSelectorFactory,
    proxyStorage: ProxyStorage,
    httpLoggingInterceptorLazy: HttpLoggingInterceptorLazy,
    interceptors: Set<@JvmSuppressWildcards KurobaOkHttpInterceptor>
  ): CoilOkHttpClient {
    deps("CoilOkHttpClient")

    return CoilOkHttpClient(
      kurobaSettings = kurobaSettings,
      normalDnsSelectorFactory = normalDnsSelectorFactory,
      dnsOverHttpsSelectorFactory = dnsOverHttpsSelectorFactory,
      proxyStorage = proxyStorage,
      httpLoggingInterceptorLazy = httpLoggingInterceptorLazy,
      interceptors = interceptors
    )
  }

  /**
   * This okHttpClient is for images/file/apk updates/ downloading, prefetching, etc.
   */
  @Provides
  @Singleton
  fun provideDownloaderOkHttpClient(
    kurobaSettings: KurobaSettings,
    normalDnsSelectorFactory: NormalDnsSelectorFactory,
    dnsOverHttpsSelectorFactory: DnsOverHttpsSelectorFactory,
    proxyStorage: ProxyStorage,
    httpLoggingInterceptorLazy: HttpLoggingInterceptorLazy,
    interceptors: Set<@JvmSuppressWildcards KurobaOkHttpInterceptor>
  ): DownloaderOkHttpClient {
    deps("DownloaderOkHttpClient")

    return DownloaderOkHttpClientImpl(
      kurobaSettings = kurobaSettings,
      normalDnsSelectorFactory = normalDnsSelectorFactory,
      dnsOverHttpsSelectorFactory = dnsOverHttpsSelectorFactory,
      proxyStorage = proxyStorage,
      httpLoggingInterceptorLazy = httpLoggingInterceptorLazy,
      interceptors = interceptors
    )
  }

  /**
   * Special OkHttpClient which is only for use inside of OkHttpInterceptors
   * */
  @Provides
  @Singleton
  fun provideOkHttpClientForInterceptors(
    kurobaSettings: KurobaSettings,
    normalDnsSelectorFactory: NormalDnsSelectorFactory,
    dnsOverHttpsSelectorFactory: DnsOverHttpsSelectorFactory,
    proxyStorage: ProxyStorage,
    httpLoggingInterceptorLazy: HttpLoggingInterceptorLazy,
  ): OkHttpClientForInterceptors {
    deps("OkHttpClientForInterceptorsImpl")
    return OkHttpClientForInterceptorsImpl(
      kurobaSettings = kurobaSettings,
      normalDnsSelectorFactory = normalDnsSelectorFactory,
      dnsOverHttpsSelectorFactory = dnsOverHttpsSelectorFactory,
      proxyStorage = proxyStorage,
      httpLoggingInterceptorLazy = httpLoggingInterceptorLazy,
    )
  }
}
